### ISSUE TITLE
Spurious irrefutable_let_patterns warning with let-chain

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#139369]
### Code

```Rust
#![feature(let_chains)]

fn max() -> usize {
    42
}

fn main() {
    if let mx = max() && mx < usize::MAX {
        // ...
    }
}
```

### Current output

```Shell
warning: leading irrefutable pattern in let chain
 --> src/main.rs:7:8
  |
7 |     if let mx = max() && mx < usize::MAX {
  |        ^^^^^^^^^^^^^^
  |
  = note: this pattern will always match
  = help: consider moving it outside of the construct
  = note: `#[warn(irrefutable_let_patterns)]` on by default
```

### Desired output

```Shell
(no warning)
```

### Rationale and extra context

This code naturally expresses "please call that function and then do something if the return value satisfies a condition". Putting the `let` binding outside the `if` would be bad as then it remains in scope after the `if`, which is not the intent. I could wrap it in an extra layer of curly braces but that increases the overall indentation level.

I think this warning should never trigger for let chains.

### Other cases

```Rust

```

### Rust Version

```Shell
current nightly
```

**Comment by kurikomoe** [comment]
another example:
```rust
#[allow(irrefutable_let_patterns)]
if let errCode = GetLastError() && errCode != ERROR_NO_MORE_ITEMS {
    let _ = dbg!(errCode);
    // Do something
}
```
Honestly, I don't want to introduce the errCode into the outer scope; I just want to keep it inside this if-else branch to avoid the possibility of using the variable elsewhere.

**Comment by traviscross** [comment]
Do you want to weaken or remove this lint in general, or do you just want it to not fire on let chains?

**Comment by joshtriplett** [comment]
I think there's a reasonable case for not applying it to let chains, specifically. In the absence of let chains, `if let name = expr() { ... }` is unnecessary and always true, and it seems reasonable to lint on that. But with let chains, it makes sense to bind something for use later in the chain.

**Comment by RalfJung** [comment]
> I seem to recall at some point that we talked about making it not fire for any subsequent let in the chain, but keeping it for the first one since that one can be moved out. 

Both of the examples posted above argue against that: moving it out extends the scope of the binding, which might not be desired.

**Comment by traviscross** [comment]
If someone wants to put up a PR to make it not fire on let chains and nominate that, I'll propose FCP merge on it.  I'd estimate that being likely to save us a step here.

**Comment by joshtriplett** [comment]
@RalfJung I'm aware, but that isn't necessarily a reason to allow it, just an argument that someone might want to write it that way.

The *downside* of allowing it for let chains in general is that it may not be intentional, and sometimes this catches errors where you meant to write a refutable pattern.

But nonetheless, I'd be :+1: for allowing it for let chains, for now, with the possibility of coming up with something more specific in the future to catch potential mistakes.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated +I-lang-radar

We talked about this in the @rust-lang/lang call today.  We decided we just wanted to rip out the `irrefutable_let_patterns` lint entirely.  Given the existing dead code and unused variable lints, we just weren't seeing a lot of incremental value in this one, and we could think of many reasons for people to write code that would hit this.  Whoever does this should ensure that `dead_code` catches e.g. `let () = () else { /* dead code */ };` and `if let () = () { .. } else { /* dead code */ }` and whatnot, and then nominate it for us.

And CC @rust-lang/clippy in the event that clippy might want to add something here.

**Comment by jnkr-ifx** [comment]
I find the `irrefutable_let_patterns` lint beneficial occasionally because I have a bad habit of writing `if let x = ...` when I'm trying to unwrap an Option. (I picked up the habit from Swift, where `if let x = ...` is syntactic sugar for `if let Some(x) = ...`).

```rust
#![allow(irrefutable_let_patterns)]

fn foo() -> Option<u32> {
    None
}

fn main() {
    if let x = foo() {
        println!("{:?}", x);
    }
}
```

This [prints `None`, with no warning](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=20da74fc83181418ad1238d6ecbfbab4).

In most cases this mistake will lead to a type error really quickly because the Option isn't unwrapped in the body of the if statement -- I had to use `{:?}` instead of `{}` to write that example -- but I can definitely remember seeing this lint fire and remind me "hey, you're not writing Swift" before, especially when I was newer to Rust.

**Comment by traviscross** [comment]
We did talk about how some aspects of the lint might be most beneficial to new users, and that's why we pinged the clippy team here, as adding a lint to help such users might better fit in their domain.

**Comment by tmandry** [comment]
The comparison to Swift is interesting; I'd forgotten about that syntax.

The specific case of `if let x = ...` being confused with `if x == ...` was mentioned, and it was brought up that this would usually result in an unused variable warning. Of course the snippet you posted doesn't hit this since `x` is used, and since debug-printing is agnostic to whether the type of the binding `x` is an option. One question is whether we think cases like this that aren't caught by unused variables or the type checker are ever "important".

That said, with example from @jnkr-ifx **I've switched back to thinking most non-chain uses of `if let` qualify as either _likely bugs_ or _needless divergence_**.

What pushed me toward removing the lint is that I don't want overly paternalistic lints that limit expressiveness needlessly. It's true that we've already experienced that with some of the "irrefutable" lints (particularly in the where clause department where we've already relaxed it some). In fact I felt some annoyance about this arising in the discussion today.

But I also don't buy the case I've heard so far for how it's nice to support `if let x = foo() { ... }` outside of a let chain. The potential for confusion with Swift feels much more concrete to me. The place where I _sort of_ buy wanting it is something like this:

```rust
if let Some(x) = foo() {
  ...
} else if let Ok(y) = bar() {
  ...
} else if let z = baz() {
  ...
}
```

And I don't know, maybe I would want to write that for aesthetic reasons, but it isn't so bad to write `else { let z = baz(); ... }`. Crucially it's also clearer to the reader, who naturally would expect any usage of `if` to be followed by some condition.

In contrast, I find the example with a let chain brought up by @kurikomoe _extremely_ motivating. Assigning to a variable and checking a condition on its value immediately is a very common pattern, and it's also common to only need that variable to exist inside the if block. Other languages have forms like this for the same reason.

Outside of let chains, my thinking now is this is one of the places where Rust should err more in the direction of Go (simple and uniform) than Haskell (elegant and inscrutable). In terms of design goals, I would argue that "clarity of purpose" and "correctness by construction" are both on the side of keeping the lint but removing it for let chains.

**Comment by traviscross** [comment]
In that case, I feel like the lint you might want is about a trivially true condition rather than about the irrefutable pattern.  These all seem about the same to me:

```rust
if let () = () {}
if let () = () && true {}
if true {}
```

**Comment by traviscross** [comment]
We're already convinced on let chains, but for posterity, here's what the most motivating case for that looks like to me:

```rust
if let x = get_thing(1)
&& x > 0
&& let y = get_thing(2)
&& y > 0
&& let z = get_thing(3)
&& z > 0
{
    ..
}
```

That is, we don't want to get `y` at all if the test on `x` fails, etc.  This factoring is saving us from having to nest many conditionals, which is really the whole point of let chains.

**Comment by RalfJung** [comment]
> These all seem about the same to me:

I don't agree for `if true`. That can be quite useful to write during debugging etc and it's annoying to have the compiler complain about it. Java does that and people resort to `if 0 == 0` etc to work around it...

**Comment by traviscross** [comment]
Yes, that's actually why I feel the same about `if let x = foo() { .. }`.  Because you start with `if let x = foo() && x > 8 && x < 16 { .. }`, and then for debugging (etc.) you want to truncate off parts of that conditional, so you'll delete characters until you maybe have `if let x = foo() { .. }`, and then the compiler will complain and want you to unwrap that body.

**Comment by joshtriplett** [comment]
Generally speaking, if I'm in the middle of debugging, and have deleted some code or added `true &&` to a conditional or similar, I don't mind getting compiler warnings, because I'm going to ignore the ones *I caused* while I'm debugging. If anything, they're a useful reminder that debugging is in progress.

**Comment by flip1995** [comment]
> And CC `@rust-lang/clippy` in the event that clippy might want to add something here.

Should this lint be removed from rustc, it could of course be "down-lifted" to Clippy. I can see it in the `complexity` (warn-by-default) group in Clippy.

**Comment by traviscross** [comment]
We discussed this in the lang call today without consensus.

**Comment by RalfJung** [comment]
But to be clear, the consensus of not emitting the lint for chains still stands, right? So we could have a PR to adjust that while the question around non-chains remains open? (The issue was originally meant to be about chains only anyway.)

**Comment by traviscross** [comment]
We didn't discuss it particularly, but yes, my estimate would be that such a PR would be an easy FCP for us.

**Comment by traviscross** [comment]
Last we discussed the point, people also seemed OK with not firing the lint on `while let`.

**Comment by traviscross** [comment]
Thinking about the Swift behavior further -- that `if let x = ..` is `if let Some(x) = ..` -- if we were to be persuaded on that basis, then I'm increasingly not sure that there'd be a consistent case for relaxing this lint so as to not fire on let chains.  Because just as we might have type confusion with,

```rust
if let x = Some("hello") {
    if format!("{x:?}") == "hello" {
        do_important_work(); //~ Silently never reached.
    }
}
```

we could equally as well have type confusion with:

```rust
if let x = Some("hello")
&& format!("{x:?}") == "hello"
{
    do_important_work(); //~ Silently never reached.
}
```

**Comment by RalfJung** [comment]
I think the fact that another language picked that convention (a convention that is rather at odds with how we handle option types in Rust) should not stop us from supporting idioms that are very natural in Rust, and that are awkward to express without firing this lint, such as the two examples at the top of this post.

**Comment by tmandry** [comment]
> So I don't really see a principled or practical case for distinguishing an irrefutable let expression from any other trivially true expression. If we want to nag people to unwrap these blocks (or at least remove the `if`), it seems to apply to both, and if we don't, that seems to apply to both as well.

The reason is that the meaning of `if true {}`and `if 0 == 0 {}` is universal and unambiguous, while the meaning of `if let x = foo() {}` is not.

**Comment by traviscross** [comment]
Because of Swift?  The angle I had been pushing there was that a `let x = ..` expression is unambiguously a trivially true expression in Rust, and so is similar in that way to other obviously true expressions.

(@scottmcm made the point that the triviality of an arbitrary let expression is a type-space calculation rather than a value-space one, and I agree that's an interesting and distinguishing point, though the common case of `let x = ..` is always as trivially true as `true` is.)

The [later point](https://github.com/rust-lang/rust/issues/139369#issuecomment-2864308360) I'd press is that if we do think `let x = ..` is ambiguous in its meaning, due to Swift, then I think that affects let chains as well, and that would be unfortunate, because the use cases here are compelling.

**Comment by traviscross** [comment]
Because of Swift, I presume.  The angle I was push there was that, regardless of how ambiguous or unambiguous we feel that `let x = ..` is, `let x = ..` and `let x = .. && true` should be equal in that respect.  It seems strange that adding `&& true` should disambiguate that case when `let x = ..` (in this context) is already a trivially true expression in Rust.

So if we write:

```rust
if let x = .. { .. } else { .. }
if let x = .. && true { .. } else { .. }
```

I feel like I'd expect the same warnings (or lack thereof) on both of those independent of our feelings about the ambiguousness of `let x = ..`.  And if we warn on that second case, then it'd feel strange to me to not also warn on

```rust
if true { .. } else { .. }
```

since the condition is "obviously" just as trivial as `let x = .. && true`.

(@scottmcm made the point that the triviality of an arbitrary let expression is a type-space calculation rather than a value-space one, and I agree that's an interesting and distinguishing point, though the common case of `let x = ..` is always as trivially true as `true` is.)

The [later point](https://github.com/rust-lang/rust/issues/139369#issuecomment-2864308360) I'd press is similarly in the spirit that the chained case may not be so different from the unchained case, if we think that `let x = ..` is ambiguous, though that would be unfortunate, because the use cases here are compelling.

**Comment by RalfJung** [comment]
I think the lint should only warn for cases that are trivially true *and where that is likely an accident*. If someone writes `if true`, that's not an accident, that's an expression of intent, and we shouldn't warn (and we currently do not warn).

The trouble with `if let x = ...` is that it can be both an expression of intent, as in the examples above, and an accident, if someone comes from Swift. But I'd rather not prioritize programmers coming from just one particular other language over useful Rust-side idioms. Most Rust programmers do not come from Swift, after all.

The argument for warning only in chains is that it helps us disambiguate those two cases: without a chain, the chance that this is intentional is much lower, since there's much less of a use for such an idiom on the Rust side, and thus the chances that this is a Swift-induced mistake is bigger and a lint is more likely to find a true mistake.

**Comment by traviscross** [comment]
I don't know.  The only time there could be type confusion is when the user isn't otherwise constraining the type sufficiently, and it's hard for me to tell what's going to be more or less likely there.  You have to work at it a bit anyway for this to be a problem, and it's not clear to me that

```rust
if let x = Some("hello")
&& format!("{x:?}") == "hello"
{
    do_important_work(); //~ Silently never reached.
}
```

would be less likely to be a problem in that respect than anything else.

Perhaps it would be a challenging lint to write, but it seems the appropriate lint for Swift users would be targeted on this, where you said `if let x = ..` and then did not later constrain the type of `x` in such a way as to make it clear you know that you're in Rust and not Swift.

**Comment by RalfJung** [comment]
I didn't say the Swift confusion is impossible in the chained case. But an irrefutable pattern is less likely to be a Swift confusion when there is a chain, mostly because there's a higher chance that someone actually wrote this on purpose without being confused.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated +I-lang-radar -P-lang-drag-2

In a short-staff portion of a call today, @tmandry and I talked about this.

Here's what I'd suggest.  Both @tmandry and I agree the use case is motivating for not linting this on let chains.  As far as I know, everyone else on lang is still on board with at least that limited step, so I'd suggest and encourage someone to put up a PR causing this lint to not fire on let chains and to nominate that for us.

Probably I'd still prefer to go further and to lint against code made dead by the use of an irrefutable pattern but not to otherwise warn-by-default lint on it, following our [original consensus](https://github.com/rust-lang/rust/issues/139369#issuecomment-2842997955).  If someone is ever inclined to put up a separate PR for that, it would get a hearing.

More broadly, I wish we had more precise tools available to us for dealing with the "somebody might be coming from C#" or "somebody might be coming from Swift" class of problem.  Not sure the solution there, but it's something to think about.

**Comment by Natural-selection1** [comment]
> How about make the lint only apply when _all_ conditions in the chain are irrefutable let bindings?

Such logic already exists in current code

https://github.com/rust-lang/rust/blob/ae12bc21d8ec76bbb753d4da168e9b08e1b09ebf/compiler/rustc_mir_build/src/thir/pattern/check_match.rs#L565-L582

**Comment by camsteffen** [comment]
@RalfJung That example has a non-let expression. I'm wondering about linting where _all_ conditions are lets with irrefutable patterns.

**Comment by camsteffen** [comment]
Yes, it seems to me "all irrefutable lets" checks all the boxes and is a more conservative change. I don't think I've seen any argument specifically for not linting all irrefutable lets.
